<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Intro to R and RStudio for Genomics: Data Wrangling and Analyses with Tidyverse</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/dc/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/dc/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/dc/favicon-16x16.png"><link rel="manifest" href="favicons/dc/site.webmanifest"><link rel="mask-icon" href="favicons/dc/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav data"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="dc-logo" alt="Data Carpentry" src="assets/images/data-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/05-dplyr.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="assets/images/data-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Intro to R and RStudio for Genomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Intro to R and RStudio for Genomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Intro to R and RStudio for Genomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 32%" class="percentage">
    32%
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: 32%" aria-valuenow="32" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/05-dplyr.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="00-introduction.html">1. Introducing R and RStudio IDE</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="01-r-basics.html">2. R Basics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="02-data-prelude.html">3. Introduction to the example dataset and file type</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        4. Data Wrangling and Analyses with Tidyverse
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#working-with-spreadsheets-tabular-data">Working with spreadsheets (tabular data)</a></li>
<li><a href="#what-is-dplyr">What is dplyr?</a></li>
<li><a href="#importing-tabular-data-into-r">Importing tabular data into R</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="06-data-visualization.html">5. Data Visualization with ggplot2</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="07-r-help.html">6. Getting help with R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="08-vectors-extra.html">7. Extra: R Basics continued - factors</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="09-basics-factors-dataframes.html">8. Extra: R Basics continued - factors and data frames</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="10-bioconductor-vcfr.html">9. Extra: Using packages from Bioconductor</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html">Glossary</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="02-data-prelude.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="06-data-visualization.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="02-data-prelude.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to the
        </a>
        <a class="chapter-link float-end" href="06-data-visualization.html" rel="next">
          Next: Data Visualization...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Data Wrangling and Analyses with Tidyverse</h1>
        <p>Last updated on 2025-08-05 |

        <a href="https://github.com/UW-Madison-DataScience/phmsci756-genomics-r-intro/edit/main/episodes/05-dplyr.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I manipulate data frames without repeating myself?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain the basic principle of tidy datasets</li>
<li>Be able to load a tabular dataset using base R functions</li>
<li>Describe what the <code>dplyr</code> package in R is used for.</li>
<li>Apply common <code>dplyr</code> functions to manipulate data in
R.</li>
<li>Employ the ‘pipe’ operator to link together a sequence of
functions.</li>
<li>Employ the ‘mutate’ function to apply other chosen functions to
existing columns and create new columns of data.</li>
<li>Employ the ‘split-apply-combine’ concept to split the data into
groups, apply analysis to each group, and combine the results.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="working-with-spreadsheets-tabular-data">Working with spreadsheets (tabular data)<a class="anchor" aria-label="anchor" href="#working-with-spreadsheets-tabular-data"></a></h2>
<hr class="half-width"><p>A substantial amount of the data we work with in genomics will be
tabular data, this is data arranged in rows and columns - also known as
spreadsheets. We could write a whole lesson on how to work with
spreadsheets effectively (<a href="https://datacarpentry.org/organization-genomics/" class="external-link">actually we
did</a>). For our purposes, we want to remind you of a few principles
before we work with our first set of example data:</p>
<p><strong>1) Keep raw data separate from analyzed data</strong></p>
<p>This is principle number one because if you can’t tell which files
are the original raw data, you risk making some serious mistakes
(e.g. drawing conclusion from data which have been manipulated in some
unknown way).</p>
<p><strong>2) Keep spreadsheet data Tidy</strong></p>
<p>The simplest principle of <strong>Tidy data</strong> is that we have
one row in our spreadsheet for each observation or sample, and one
column for every variable that we measure or report on. As simple as
this sounds, it’s very easily violated. Most data scientists agree that
significant amounts of their time is spent tidying data for analysis.
Read more about data organization in <a href="https://datacarpentry.org/organization-genomics/" class="external-link">our lesson</a>
and in <a href="https://www.jstatsoft.org/article/view/v059i10" class="external-link">this
paper</a>.</p>
<p><strong>3) Trust but verify</strong></p>
<p>Finally, while you don’t need to be paranoid about data, you should
have a plan for how you will prepare it for analysis. <strong>This a
focus of this lesson.</strong> You probably already have a lot of
intuition, expectations, assumptions about your data - the range of
values you expect, how many values should have been recorded, etc. Of
course, as the data get larger our human ability to keep track will
start to fail (and yes, it can fail for small data sets too). R will
help you to examine your data so that you can have greater confidence in
your analysis, and its reproducibility.</p>
<div id="tip-keep-your-raw-data-separate" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-keep-your-raw-data-separate" class="callout-inner">
<h3 class="callout-title">Tip: Keep your raw data separate</h3>
<div class="callout-content">
<p>When you work with data in R, you are not changing the original file
you loaded that data from. This is different than (for example) working
with a spreadsheet program where changing the value of the cell leaves
you one “save”-click away from overwriting the original file. You have
to purposely use a writing function (e.g. <code>write.csv()</code>) to
save data loaded into R. In that case, be sure to save the manipulated
data into a new file. More on this later in the lesson.</p>
</div>
</div>
</div>
<p>Base R (without additional packages) has a way of subsetting using
brakets, which is handy, but it can be cumbersome and difficult to read,
especially for complicated operations.</p>
<p>Luckily, the <a href="https://cran.r-project.org/package=dplyr" class="external-link"><code>dplyr</code></a>
package provides a number of very useful functions for manipulating data
frames (aka spreadsheets or tables of data) in a way that will reduce
repetition, reduce the probability of making errors, and probably even
save you some typing. As an added bonus, you might even find the
<code>dplyr</code> grammar easier to read.</p>
<p>Here we’re going to cover some of the most commonly used functions as
well as using pipes (<code>%&gt;%</code>) to combine them:</p>
<ol style="list-style-type: decimal"><li><code>glimpse()</code></li>
<li><code>select()</code></li>
<li><code>filter()</code></li>
<li><code>group_by()</code></li>
<li><code>summarize()</code></li>
<li><code>mutate()</code></li>
<li>
<code>inner_join()</code>, <code>full_join()</code>,
<code>left_join()</code>, <code>right_join()</code>
</li>
<li>Extra - <code>pivot_longer</code> and <code>pivot_wider</code>
</li>
</ol><p>Packages in R are sets of additional functions that let you do more
stuff in R. The functions we’ve been using, like <code>str()</code>,
come built into R; packages give you access to more functions. You need
to install a package and then load it to be able to use it.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span> <span class="co">## installs dplyr package</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"tidyr"</span><span class="op">)</span> <span class="co">## installs tidyr package</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span> <span class="co">## installs ggplot2 package</span></span>
<span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"readr"</span><span class="op">)</span> <span class="co">## install readr package</span></span></code></pre>
</div>
<div id="tip-installing-packages" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-installing-packages" class="callout-inner">
<h3 class="callout-title">Tip: Installing packages</h3>
<div class="callout-content">
<p>It may be temping to install the <code>tidyverse</code> package, as
it contains many useful collection of packages for this lesson and
beyond. However, when teaching or following this lesson, we advise that
participants install <code>dplyr</code>, <code>readr</code>,
<code>ggplot2</code>, and <code>tidyr</code> individually as shown
above. Otherwise, a substaial amount of the lesson will be spend waiting
for the installation to complete.</p>
</div>
</div>
</div>
<p>You might get asked to choose a CRAN mirror – this is asking you to
choose a site to download the package from. The choice doesn’t matter
too much; I’d recommend choosing the RStudio mirror.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span>          <span class="co">## loads in dplyr package to use</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"tidyr"</span><span class="op">)</span>          <span class="co">## loads in tidyr package to use</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span>          <span class="co">## loads in ggplot2 package to use</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"readr"</span><span class="op">)</span>          <span class="co">## load in readr package to use</span></span></code></pre>
</div>
<p>You only need to install a package once per computer, but you need to
load it every time you open a new R session and want to use that
package.</p>
</section><section><h2 class="section-heading" id="what-is-dplyr">What is dplyr?<a class="anchor" aria-label="anchor" href="#what-is-dplyr"></a></h2>
<hr class="half-width"><p>The package <code>dplyr</code> is a fairly new (2014) package that
tries to provide easy tools for the most common data manipulation tasks.
This package is also included in the <a href="https://www.tidyverse.org/" class="external-link"><code>tidyverse</code> package</a>,
which is a collection of eight different packages (<code>dplyr</code>,
<code>ggplot2</code>, <code>tibble</code>, <code>tidyr</code>,
<code>readr</code>, <code>purrr</code>, <code>stringr</code>, and
<code>forcats</code>). It is built to work directly with data frames.
The thinking behind it was largely inspired by the package
<code>plyr</code> which has been in use for some time but suffered from
being slow in some cases.<code>dplyr</code> addresses this by porting
much of the computation to C++. An additional feature is the ability to
work with data stored directly in an external database. The benefits of
doing this are that the data can be managed natively in a relational
database, queries can be conducted on that database, and only the
results of the query returned.</p>
<p>This addresses a common problem with R in that all operations are
conducted in memory and thus the amount of data you can work with is
limited by available memory. The database connections essentially remove
that limitation in that you can have a database that is over 100s of GB,
conduct queries on it directly and pull back just what you need for
analysis in R.</p>
</section><section><h2 class="section-heading" id="importing-tabular-data-into-r">Importing tabular data into R<a class="anchor" aria-label="anchor" href="#importing-tabular-data-into-r"></a></h2>
<hr class="half-width"><p>There are several ways to import data into R. We will start loading
our data using the tidyverse package called <code>readr</code> and a
function called <code>read_csv()</code></p>
<div id="exercise-review-the-arguments-of-the-read_csv-function" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-review-the-arguments-of-the-read_csv-function" class="callout-inner">
<h3 class="callout-title">Exercise: Review the arguments of the
<code>read_csv()</code> function</h3>
<div class="callout-content">
<p><strong>Before using the <code>read_csv()</code> function, use R’s
help feature to answer the following questions</strong>.</p>
<p><em>Hint</em>: Entering ‘?’ before the function name and then running
that line will bring up the help documentation. Also,
<code>read_csv()</code> is part of a family of functions for reading in
data called <code>read_delim()</code> so you will need to look for the
help page for <code>read_delim()</code> instead. When reading this
particular help be careful to pay attention to the ‘read_csv’ expression
under the ‘Usage’ heading. Other answers will be in the ‘Arguments’
heading.</p>
<ol style="list-style-type: upper-alpha"><li><p>What is the default parameter for ‘col_names’ in the
<code>read_csv()</code> function?</p></li>
<li><p>What argument would you have to change to read a file that was
delimited by semicolons (;) rather than commas?</p></li>
<li><p>What argument would you have to change to read skip commented
lines (starting with #) at the beginning of a file (like our VCF file)?
Hint: There are a couple of different possible answers to this
question.</p></li>
<li><p>What argument would you have to change to read in only the first
10,000 rows of a very large file?</p></li>
</ol></div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol style="list-style-type: upper-alpha"><li><p>The <code>read_csv()</code> function has the argument ‘col_names’
set to TRUE by default, this means the function always assumes the first
row is header information, (i.e. column names)</p></li>
<li><p>The <code>read_csv()</code> function has the argument ‘delim’
which allows you to change the delimiter (aka separator) between
columns. The function assumes commas are used as delimiters, as you
would expect. Changing this parameter (e.g. <code>delim=";"</code>)
would now interpret semicolons as delimiters.</p></li>
<li><p>To skip commented lines at the beginning of a file, there are a
couple options. If the number of lines is consistent, we can use the
<code>skip</code> argument, for example, <code>skip = 3</code> will skip
the first 3 lines of the file and then start reading in the header and
data starting on line</p></li>
</ol><ol start="4" style="list-style-type: decimal"><li>If the commented lines use a consistent delimeter (like #) we can
instead use the <code>comment</code> argument, which skips any
information after the charcter given. In this example
<code>comment = '#'</code> will ignore lines starting with the
hashtag/pound symbol. Note if you use both, skip will be executed first
and then comment.</li>
</ol><ol start="4" style="list-style-type: upper-alpha"><li>You can set <code>n_max</code> to a numeric value
(e.g. <code>n_max=10000</code>) to choose how many rows of a file you
read in. This may be useful for very large files where not all the data
is needed to test some data cleaning steps you are applying.</li>
</ol><p>Hopefully, this exercise gets you thinking about using the provided
help documentation in R. There are many arguments that exist, but which
we wont have time to cover. Look here to get familiar with functions you
use frequently, you may be surprised at what you find they can do.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="loading--csv-files-in-tidy-style">Loading .csv files in tidy style<a class="anchor" aria-label="anchor" href="#loading--csv-files-in-tidy-style"></a></h3>
<p>Now let’s load our vcf .csv file using <code>read_csv()</code>:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 801 Columns: 29
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (7): sample_id, CHROM, REF, ALT, DP4, Indiv, gt_GT_alleles
dbl (16): POS, QUAL, IDV, IMF, DP, VDB, RPB, MQB, BQB, MQSB, SGB, MQ0F, AC, ...
num  (1): gt_PL
lgl  (5): ID, FILTER, INDEL, ICB, HOB

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="taking-a-quick-look-at-data-frames">Taking a quick look at data frames<a class="anchor" aria-label="anchor" href="#taking-a-quick-look-at-data-frames"></a></h3>
<p>Similar to <code>str()</code>, which comes built into R,
<code>glimpse()</code> is a <code>dplyr</code> function that (as the
name suggests) gives a glimpse of the data frame.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 801
Columns: 29
$ sample_id     &lt;chr&gt; "SRR2584863", "SRR2584863", "SRR2584863", "SRR2584863", …
$ CHROM         &lt;chr&gt; "CP000819.1", "CP000819.1", "CP000819.1", "CP000819.1", …
$ POS           &lt;dbl&gt; 9972, 263235, 281923, 433359, 473901, 648692, 1331794, 1…
$ ID            &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ REF           &lt;chr&gt; "T", "G", "G", "CTTTTTTT", "CCGC", "C", "C", "G", "ACAGC…
$ ALT           &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…
$ QUAL          &lt;dbl&gt; 91.0000, 85.0000, 217.0000, 64.0000, 228.0000, 210.0000,…
$ FILTER        &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ INDEL         &lt;lgl&gt; FALSE, FALSE, FALSE, TRUE, TRUE, FALSE, FALSE, FALSE, TR…
$ IDV           &lt;dbl&gt; NA, NA, NA, 12, 9, NA, NA, NA, 2, 7, NA, NA, NA, NA, NA,…
$ IMF           &lt;dbl&gt; NA, NA, NA, 1.000000, 0.900000, NA, NA, NA, 0.666667, 1.…
$ DP            &lt;dbl&gt; 4, 6, 10, 12, 10, 10, 8, 11, 3, 7, 9, 20, 12, 19, 15, 10…
$ VDB           &lt;dbl&gt; 0.0257451, 0.0961330, 0.7740830, 0.4777040, 0.6595050, 0…
$ RPB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.900802, …
$ MQB           &lt;dbl&gt; NA, 1.0000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.1501340…
$ BQB           &lt;dbl&gt; NA, 1.000000, NA, NA, NA, NA, NA, NA, NA, NA, 0.750668, …
$ MQSB          &lt;dbl&gt; NA, NA, 0.974597, 1.000000, 0.916482, 0.916482, 0.900802…
$ SGB           &lt;dbl&gt; -0.556411, -0.590765, -0.662043, -0.676189, -0.662043, -…
$ MQ0F          &lt;dbl&gt; 0.000000, 0.166667, 0.000000, 0.000000, 0.000000, 0.0000…
$ ICB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ HOB           &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ AC            &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ AN            &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ DP4           &lt;chr&gt; "0,0,0,4", "0,1,0,5", "0,0,4,5", "0,1,3,8", "1,0,2,7", "…
$ MQ            &lt;dbl&gt; 60, 33, 60, 60, 60, 60, 60, 60, 60, 60, 25, 60, 10, 60, …
$ Indiv         &lt;chr&gt; "/home/dcuser/dc_workshop/results/bam/SRR2584863.aligned…
$ gt_PL         &lt;dbl&gt; 1210, 1120, 2470, 910, 2550, 2400, 2080, 2550, 11128, 19…
$ gt_GT         &lt;dbl&gt; 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
$ gt_GT_alleles &lt;chr&gt; "G", "T", "T", "CTTTTTTTT", "CCGCGC", "T", "A", "A", "AC…</code></pre>
</div>
<p>In the above output, we can already gather some information about
<code>variants</code>, such as the number of rows and columns, column
names, type of vector in the columns, and the first few entries of each
column. Although what we see is similar to outputs of
<code>str()</code>, this method gives a cleaner visual output.</p>
</div>
<div class="section level3">
<h3 id="selecting-columns-and-filtering-rows">Selecting columns and filtering rows<a class="anchor" aria-label="anchor" href="#selecting-columns-and-filtering-rows"></a></h3>
<p>To select columns of a data frame, use <code>select()</code>. The
first argument to this function is the data frame
(<code>variants</code>), and the subsequent arguments are the columns to
keep.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 4
   sample_id  REF                              ALT                            DP
   &lt;chr&gt;      &lt;chr&gt;                            &lt;chr&gt;                       &lt;dbl&gt;
 1 SRR2584863 T                                G                               4
 2 SRR2584863 G                                T                               6
 3 SRR2584863 G                                T                              10
 4 SRR2584863 CTTTTTTT                         CTTTTTTTT                      12
 5 SRR2584863 CCGC                             CCGCGC                         10
 6 SRR2584863 C                                T                              10
 7 SRR2584863 C                                A                               8
 8 SRR2584863 G                                A                              11
 9 SRR2584863 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCC…     3
10 SRR2584863 AT                               ATT                             7
# ℹ 791 more rows</code></pre>
</div>
<p>To select all columns <em>except</em> certain ones, put a “-” in
front of the variable to exclude it.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="op">-</span><span class="va">CHROM</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 28
   sample_id      POS ID    REF      ALT    QUAL FILTER INDEL   IDV    IMF    DP
   &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt;    &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
 1 SRR2584863    9972 NA    T        G        91 NA     FALSE    NA NA         4
 2 SRR2584863  263235 NA    G        T        85 NA     FALSE    NA NA         6
 3 SRR2584863  281923 NA    G        T       217 NA     FALSE    NA NA        10
 4 SRR2584863  433359 NA    CTTTTTTT CTTT…    64 NA     TRUE     12  1        12
 5 SRR2584863  473901 NA    CCGC     CCGC…   228 NA     TRUE      9  0.9      10
 6 SRR2584863  648692 NA    C        T       210 NA     FALSE    NA NA        10
 7 SRR2584863 1331794 NA    C        A       178 NA     FALSE    NA NA         8
 8 SRR2584863 1733343 NA    G        A       225 NA     FALSE    NA NA        11
 9 SRR2584863 2103887 NA    ACAGCCA… ACAG…    56 NA     TRUE      2  0.667     3
10 SRR2584863 2333538 NA    AT       ATT     167 NA     TRUE      7  1         7
# ℹ 791 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p><code>dplyr</code> also provides useful functions to select columns
based on their names. For instance, <code>ends_with()</code> allows you
to select columns that ends with specific letters. For instance, if you
wanted to select columns that end with the letter “B”:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="fu">ends_with</span><span class="op">(</span><span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 8
      VDB   RPB   MQB   BQB   MQSB    SGB ICB   HOB
    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;lgl&gt;
 1 0.0257    NA    NA    NA NA     -0.556 NA    NA
 2 0.0961     1     1     1 NA     -0.591 NA    NA
 3 0.774     NA    NA    NA  0.975 -0.662 NA    NA
 4 0.478     NA    NA    NA  1     -0.676 NA    NA
 5 0.660     NA    NA    NA  0.916 -0.662 NA    NA
 6 0.268     NA    NA    NA  0.916 -0.670 NA    NA
 7 0.624     NA    NA    NA  0.901 -0.651 NA    NA
 8 0.992     NA    NA    NA  1.01  -0.670 NA    NA
 9 0.902     NA    NA    NA  1     -0.454 NA    NA
10 0.568     NA    NA    NA  1.01  -0.617 NA    NA
# ℹ 791 more rows</code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Create a table that contains all the columns with the letter “i” and
column “POS”, without columns “Indiv” and “FILTER”. Hint: look at for a
function called <code>contains()</code>, which can be found in the help
documentation for ends with we just covered (<code>?ends_with</code>).
Note that contains() is not case sensistive.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, we select "POS" and all columns with letter "i". This will contain columns Indiv and FILTER. </span></span>
<span><span class="va">variants_subset</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="va">POS</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"i"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Next, we remove columns Indiv and FILTER</span></span>
<span><span class="va">variants_result</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">variants_subset</span>, <span class="op">-</span><span class="va">Indiv</span>, <span class="op">-</span><span class="va">FILTER</span><span class="op">)</span></span>
<span><span class="va">variants_result</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 7
       POS sample_id  ID    INDEL   IDV    IMF ICB
     &lt;dbl&gt; &lt;chr&gt;      &lt;lgl&gt; &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;lgl&gt;
 1    9972 SRR2584863 NA    FALSE    NA NA     NA
 2  263235 SRR2584863 NA    FALSE    NA NA     NA
 3  281923 SRR2584863 NA    FALSE    NA NA     NA
 4  433359 SRR2584863 NA    TRUE     12  1     NA
 5  473901 SRR2584863 NA    TRUE      9  0.9   NA
 6  648692 SRR2584863 NA    FALSE    NA NA     NA
 7 1331794 SRR2584863 NA    FALSE    NA NA     NA
 8 1733343 SRR2584863 NA    FALSE    NA NA     NA
 9 2103887 SRR2584863 NA    TRUE      2  0.667 NA
10 2333538 SRR2584863 NA    TRUE      7  1     NA
# ℹ 791 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<em> (continued)</em>
</h3>
<div class="callout-content">
<p>We can also get to <code>variants_result</code> in one line of
code:</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Alternative solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_result</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">variants</span>, <span class="va">POS</span>, <span class="fu">contains</span><span class="op">(</span><span class="st">"i"</span><span class="op">)</span>, <span class="op">-</span><span class="va">Indiv</span>, <span class="op">-</span><span class="va">FILTER</span><span class="op">)</span></span>
<span><span class="va">variants_result</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 7
       POS sample_id  ID    INDEL   IDV    IMF ICB
     &lt;dbl&gt; &lt;chr&gt;      &lt;lgl&gt; &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;lgl&gt;
 1    9972 SRR2584863 NA    FALSE    NA NA     NA
 2  263235 SRR2584863 NA    FALSE    NA NA     NA
 3  281923 SRR2584863 NA    FALSE    NA NA     NA
 4  433359 SRR2584863 NA    TRUE     12  1     NA
 5  473901 SRR2584863 NA    TRUE      9  0.9   NA
 6  648692 SRR2584863 NA    FALSE    NA NA     NA
 7 1331794 SRR2584863 NA    FALSE    NA NA     NA
 8 1733343 SRR2584863 NA    FALSE    NA NA     NA
 9 2103887 SRR2584863 NA    TRUE      2  0.667 NA
10 2333538 SRR2584863 NA    TRUE      7  1     NA
# ℹ 791 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>To choose rows, use <code>filter()</code>:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 29
   sample_id  CHROM        POS ID    REF   ALT    QUAL FILTER INDEL   IDV    IMF
   &lt;chr&gt;      &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 SRR2584863 CP000819… 9.97e3 NA    T     G        91 NA     FALSE    NA NA
 2 SRR2584863 CP000819… 2.63e5 NA    G     T        85 NA     FALSE    NA NA
 3 SRR2584863 CP000819… 2.82e5 NA    G     T       217 NA     FALSE    NA NA
 4 SRR2584863 CP000819… 4.33e5 NA    CTTT… CTTT…    64 NA     TRUE     12  1
 5 SRR2584863 CP000819… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9  0.9
 6 SRR2584863 CP000819… 6.49e5 NA    C     T       210 NA     FALSE    NA NA
 7 SRR2584863 CP000819… 1.33e6 NA    C     A       178 NA     FALSE    NA NA
 8 SRR2584863 CP000819… 1.73e6 NA    G     A       225 NA     FALSE    NA NA
 9 SRR2584863 CP000819… 2.10e6 NA    ACAG… ACAG…    56 NA     TRUE      2  0.667
10 SRR2584863 CP000819… 2.33e6 NA    AT    ATT     167 NA     TRUE      7  1
# ℹ 15 more rows
# ℹ 18 more variables: DP &lt;dbl&gt;, VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;,
#   MQSB &lt;dbl&gt;, SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;,
#   AN &lt;dbl&gt;, DP4 &lt;chr&gt;, MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;,
#   gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p><code>filter()</code> will keep all the rows that match the
conditions that are provided. Here are a few examples:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows for which the reference genome has T or G</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">REF</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"T"</span>, <span class="st">"G"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 340 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 9.97e3 NA    T     G      91   NA     FALSE    NA    NA     4
 2 SRR25848… CP00… 2.63e5 NA    G     T      85   NA     FALSE    NA    NA     6
 3 SRR25848… CP00… 2.82e5 NA    G     T     217   NA     FALSE    NA    NA    10
 4 SRR25848… CP00… 1.73e6 NA    G     A     225   NA     FALSE    NA    NA    11
 5 SRR25848… CP00… 2.62e6 NA    G     T      31.9 NA     FALSE    NA    NA    12
 6 SRR25848… CP00… 3.00e6 NA    G     A     225   NA     FALSE    NA    NA    15
 7 SRR25848… CP00… 3.91e6 NA    G     T     225   NA     FALSE    NA    NA    10
 8 SRR25848… CP00… 9.97e3 NA    T     G     214   NA     FALSE    NA    NA    10
 9 SRR25848… CP00… 1.06e4 NA    G     A     225   NA     FALSE    NA    NA    11
10 SRR25848… CP00… 6.40e4 NA    G     A     225   NA     FALSE    NA    NA    18
# ℹ 330 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows that have TRUE in the column INDEL</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">INDEL</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 101 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 4.33e5 NA    CTTT… CTTT…  64   NA     TRUE     12 1        12
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC… 228   NA     TRUE      9 0.9      10
 3 SRR25848… CP00… 2.10e6 NA    ACAG… ACAG…  56   NA     TRUE      2 0.667     3
 4 SRR25848… CP00… 2.33e6 NA    AT    ATT   167   NA     TRUE      7 1         7
 5 SRR25848… CP00… 3.90e6 NA    A     AC     43.4 NA     TRUE      2 1         2
 6 SRR25848… CP00… 4.43e6 NA    TGG   T     228   NA     TRUE     10 1        10
 7 SRR25848… CP00… 1.48e5 NA    AGGGG AGGG… 122   NA     TRUE      8 1         8
 8 SRR25848… CP00… 1.58e5 NA    GTTT… GTTT…  19.5 NA     TRUE      6 1         6
 9 SRR25848… CP00… 1.73e5 NA    CAA   CA    180   NA     TRUE     11 1        11
10 SRR25848… CP00… 1.75e5 NA    GAA   GA    194   NA     TRUE     10 1        10
# ℹ 91 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows that don't have missing data in the IDV column</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">IDV</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 101 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 4.33e5 NA    CTTT… CTTT…  64   NA     TRUE     12 1        12
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC… 228   NA     TRUE      9 0.9      10
 3 SRR25848… CP00… 2.10e6 NA    ACAG… ACAG…  56   NA     TRUE      2 0.667     3
 4 SRR25848… CP00… 2.33e6 NA    AT    ATT   167   NA     TRUE      7 1         7
 5 SRR25848… CP00… 3.90e6 NA    A     AC     43.4 NA     TRUE      2 1         2
 6 SRR25848… CP00… 4.43e6 NA    TGG   T     228   NA     TRUE     10 1        10
 7 SRR25848… CP00… 1.48e5 NA    AGGGG AGGG… 122   NA     TRUE      8 1         8
 8 SRR25848… CP00… 1.58e5 NA    GTTT… GTTT…  19.5 NA     TRUE      6 1         6
 9 SRR25848… CP00… 1.73e5 NA    CAA   CA    180   NA     TRUE     11 1        11
10 SRR25848… CP00… 1.75e5 NA    GAA   GA    194   NA     TRUE     10 1        10
# ℹ 91 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p>We have a column titled “QUAL”. This is a Phred-scaled confidence
score that a polymorphism exists at this position given the sequencing
data. Lower QUAL scores indicate low probability of a polymorphism
existing at that site. <code>filter()</code> can be useful for selecting
mutations that have a QUAL score above a certain threshold:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rows with QUAL values greater than or equal to 100</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">QUAL</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 666 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 2.82e5 NA    G     T       217 NA     FALSE    NA  NA      10
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9   0.9    10
 3 SRR25848… CP00… 6.49e5 NA    C     T       210 NA     FALSE    NA  NA      10
 4 SRR25848… CP00… 1.33e6 NA    C     A       178 NA     FALSE    NA  NA       8
 5 SRR25848… CP00… 1.73e6 NA    G     A       225 NA     FALSE    NA  NA      11
 6 SRR25848… CP00… 2.33e6 NA    AT    ATT     167 NA     TRUE      7   1       7
 7 SRR25848… CP00… 2.41e6 NA    A     C       104 NA     FALSE    NA  NA       9
 8 SRR25848… CP00… 2.45e6 NA    A     C       225 NA     FALSE    NA  NA      20
 9 SRR25848… CP00… 2.67e6 NA    A     T       225 NA     FALSE    NA  NA      19
10 SRR25848… CP00… 3.00e6 NA    G     A       225 NA     FALSE    NA  NA      15
# ℹ 656 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<p><code>filter()</code> allows you to combine multiple conditions. You
can separate them using a <code>,</code> as arguments to the function,
they will be combined using the <code>&amp;</code> (AND) logical
operator. If you need to use the <code>|</code> (OR) logical operator,
you can specify it explicitly:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this is equivalent to:</span></span>
<span><span class="co">#   filter(variants, sample_id == "SRR2584863" &amp; QUAL &gt;= 100)</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>, <span class="va">QUAL</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 19 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 2.82e5 NA    G     T       217 NA     FALSE    NA  NA      10
 2 SRR25848… CP00… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9   0.9    10
 3 SRR25848… CP00… 6.49e5 NA    C     T       210 NA     FALSE    NA  NA      10
 4 SRR25848… CP00… 1.33e6 NA    C     A       178 NA     FALSE    NA  NA       8
 5 SRR25848… CP00… 1.73e6 NA    G     A       225 NA     FALSE    NA  NA      11
 6 SRR25848… CP00… 2.33e6 NA    AT    ATT     167 NA     TRUE      7   1       7
 7 SRR25848… CP00… 2.41e6 NA    A     C       104 NA     FALSE    NA  NA       9
 8 SRR25848… CP00… 2.45e6 NA    A     C       225 NA     FALSE    NA  NA      20
 9 SRR25848… CP00… 2.67e6 NA    A     T       225 NA     FALSE    NA  NA      19
10 SRR25848… CP00… 3.00e6 NA    G     A       225 NA     FALSE    NA  NA      15
11 SRR25848… CP00… 3.34e6 NA    A     C       211 NA     FALSE    NA  NA      10
12 SRR25848… CP00… 3.40e6 NA    C     A       225 NA     FALSE    NA  NA      14
13 SRR25848… CP00… 3.48e6 NA    A     G       200 NA     FALSE    NA  NA       9
14 SRR25848… CP00… 3.49e6 NA    A     C       225 NA     FALSE    NA  NA      13
15 SRR25848… CP00… 3.91e6 NA    G     T       225 NA     FALSE    NA  NA      10
16 SRR25848… CP00… 4.10e6 NA    A     G       225 NA     FALSE    NA  NA      16
17 SRR25848… CP00… 4.20e6 NA    A     C       225 NA     FALSE    NA  NA      11
18 SRR25848… CP00… 4.43e6 NA    TGG   T       228 NA     TRUE     10   1      10
19 SRR25848… CP00… 4.62e6 NA    A     C       185 NA     FALSE    NA  NA       9
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># using `|` logical operator</span></span>
<span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span>, <span class="op">(</span><span class="va">MQ</span> <span class="op">&gt;=</span> <span class="fl">50</span> <span class="op">|</span> <span class="va">QUAL</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 23 × 29
   sample_id  CHROM        POS ID    REF   ALT    QUAL FILTER INDEL   IDV    IMF
   &lt;chr&gt;      &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 SRR2584863 CP000819… 9.97e3 NA    T     G        91 NA     FALSE    NA NA
 2 SRR2584863 CP000819… 2.82e5 NA    G     T       217 NA     FALSE    NA NA
 3 SRR2584863 CP000819… 4.33e5 NA    CTTT… CTTT…    64 NA     TRUE     12  1
 4 SRR2584863 CP000819… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9  0.9
 5 SRR2584863 CP000819… 6.49e5 NA    C     T       210 NA     FALSE    NA NA
 6 SRR2584863 CP000819… 1.33e6 NA    C     A       178 NA     FALSE    NA NA
 7 SRR2584863 CP000819… 1.73e6 NA    G     A       225 NA     FALSE    NA NA
 8 SRR2584863 CP000819… 2.10e6 NA    ACAG… ACAG…    56 NA     TRUE      2  0.667
 9 SRR2584863 CP000819… 2.33e6 NA    AT    ATT     167 NA     TRUE      7  1
10 SRR2584863 CP000819… 2.41e6 NA    A     C       104 NA     FALSE    NA NA
# ℹ 13 more rows
# ℹ 18 more variables: DP &lt;dbl&gt;, VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;,
#   MQSB &lt;dbl&gt;, SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;,
#   AN &lt;dbl&gt;, DP4 &lt;chr&gt;, MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;,
#   gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Select all the mutations that occurred between the positions 1e6 (one
million) and 2e6 (inclusive) that have a QUAL greater than 200, and
exclude INDEL mutations. Hint: to flip logical values such as TRUE to a
FALSE, we can use to negation symbol “!”. (eg. !TRUE == FALSE).</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">POS</span> <span class="op">&gt;=</span> <span class="fl">1e6</span> <span class="op">&amp;</span> <span class="va">POS</span> <span class="op">&lt;=</span> <span class="fl">2e6</span>, <span class="va">QUAL</span> <span class="op">&gt;</span> <span class="fl">200</span>, <span class="op">!</span><span class="va">INDEL</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 77 × 29
   sample_id CHROM    POS ID    REF   ALT    QUAL FILTER INDEL   IDV   IMF    DP
   &lt;chr&gt;     &lt;chr&gt;  &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SRR25848… CP00… 1.73e6 NA    G     A       225 NA     FALSE    NA    NA    11
 2 SRR25848… CP00… 1.00e6 NA    A     G       225 NA     FALSE    NA    NA    15
 3 SRR25848… CP00… 1.02e6 NA    A     G       225 NA     FALSE    NA    NA    12
 4 SRR25848… CP00… 1.06e6 NA    C     T       225 NA     FALSE    NA    NA    17
 5 SRR25848… CP00… 1.06e6 NA    A     G       206 NA     FALSE    NA    NA     9
 6 SRR25848… CP00… 1.07e6 NA    G     T       225 NA     FALSE    NA    NA    11
 7 SRR25848… CP00… 1.07e6 NA    T     C       225 NA     FALSE    NA    NA    12
 8 SRR25848… CP00… 1.10e6 NA    C     T       225 NA     FALSE    NA    NA    15
 9 SRR25848… CP00… 1.11e6 NA    C     T       212 NA     FALSE    NA    NA     9
10 SRR25848… CP00… 1.11e6 NA    A     G       225 NA     FALSE    NA    NA    14
# ℹ 67 more rows
# ℹ 17 more variables: VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;, MQSB &lt;dbl&gt;,
#   SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;, AN &lt;dbl&gt;, DP4 &lt;chr&gt;,
#   MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;, gt_GT_alleles &lt;chr&gt;</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="pipes">Pipes<a class="anchor" aria-label="anchor" href="#pipes"></a></h3>
<p>But what if you wanted to select and filter? We can do this with
pipes. Pipes let you take the output of one function and send it
directly to the next, which is useful when you need to many things to
the same data set. It was possible to do this before pipes were added to
R, but it was much messier and more difficult. Pipes in R look like
<code>%&gt;%</code> and are made available via the <code>magrittr</code>
package, which is installed as part of <code>dplyr</code>. If you use
RStudio, you can type the pipe with <kbd>Ctrl</kbd> + <kbd>Shift</kbd> +
<kbd>M</kbd> if you’re using a PC, or <kbd>Cmd</kbd> + <kbd>Shift</kbd>
+ <kbd>M</kbd> if you’re using a Mac.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 3
   REF                              ALT                                       DP
   &lt;chr&gt;                            &lt;chr&gt;                                  &lt;dbl&gt;
 1 T                                G                                          4
 2 G                                T                                          6
 3 G                                T                                         10
 4 CTTTTTTT                         CTTTTTTTT                                 12
 5 CCGC                             CCGCGC                                    10
 6 C                                T                                         10
 7 C                                A                                          8
 8 G                                A                                         11
 9 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGC…     3
10 AT                               ATT                                        7
# ℹ 15 more rows</code></pre>
</div>
<p>In the above code, we use the pipe to send the <code>variants</code>
data set first through <code>filter()</code>, to keep rows where
<code>sample_id</code> matches a particular sample, and then through
<code>select()</code> to keep only the <code>REF</code>,
<code>ALT</code>, and <code>DP</code> columns. Since <code>%&gt;%</code>
takes the object on its left and passes it as the first argument to the
function on its right, we don’t need to explicitly include the data
frame as an argument to the <code>filter()</code> and
<code>select()</code> functions any more.</p>
<div id="same-code-with-no-pipes" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="same-code-with-no-pipes" class="callout-inner">
<h3 class="callout-title">Same code with no pipes</h3>
<div class="callout-content">
<p>Without pipes the code would look like the following.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu">select</span><span class="op">(</span><span class="fu">filter</span><span class="op">(</span><span class="va">variants</span>, <span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 3
   REF                              ALT                                       DP
   &lt;chr&gt;                            &lt;chr&gt;                                  &lt;dbl&gt;
 1 T                                G                                          4
 2 G                                T                                          6
 3 G                                T                                         10
 4 CTTTTTTT                         CTTTTTTTT                                 12
 5 CCGC                             CCGCGC                                    10
 6 C                                T                                         10
 7 C                                A                                          8
 8 G                                A                                         11
 9 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGC…     3
10 AT                               ATT                                        7
# ℹ 15 more rows</code></pre>
</div>
<p>In this code we do the filter first and then wrap the
<code>select</code> function around it to use the output of the
<code>filter</code> command as the input fo the <code>select</code>
function. While both are valid, using pipes is considered more readable
as it displays the functions in top down order, instead of inside out
order.</p>
</div>
</div>
</div>
<p>Some may find it helpful to read the pipe like the word “then”. For
instance, in the above example, we took the data frame
<code>variants</code>, <em>then</em> we <code>filter</code>ed for rows
where <code>sample_id</code> was SRR2584863, <em>then</em> we
<code>select</code>ed the <code>REF</code>, <code>ALT</code>, and
<code>DP</code> columns. The <strong><code>dplyr</code></strong>
functions by themselves are somewhat simple, but by combining them into
linear workflows with the pipe, we can accomplish more complex
manipulations of data frames.</p>
<p>If we want to create a new object with this smaller version of the
data we can do so by assigning it a new name:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span> <span class="op">&lt;-</span> <span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">DP</span><span class="op">)</span></span></code></pre>
</div>
<p>This new object includes all of the data from this sample. Let’s look
at just the first six rows to confirm it’s what we want:</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 3
   REF                              ALT                                       DP
   &lt;chr&gt;                            &lt;chr&gt;                                  &lt;dbl&gt;
 1 T                                G                                          4
 2 G                                T                                          6
 3 G                                T                                         10
 4 CTTTTTTT                         CTTTTTTTT                                 12
 5 CCGC                             CCGCGC                                    10
 6 C                                T                                         10
 7 C                                A                                          8
 8 G                                A                                         11
 9 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGC…     3
10 AT                               ATT                                        7
# ℹ 15 more rows</code></pre>
</div>
<p>We can use the <code>head()</code> and <code>tail()</code> functions
to look at the first or last six rows. There is also a more versitle
tidyverse function <code>slice()</code>, that allows users to specify a
range to view:</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span> <span class="op">%&gt;%</span> <span class="fu">head</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 3
  REF      ALT          DP
  &lt;chr&gt;    &lt;chr&gt;     &lt;dbl&gt;
1 T        G             4
2 G        T             6
3 G        T            10
4 CTTTTTTT CTTTTTTTT    12
5 CCGC     CCGCGC       10
6 C        T            10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span> <span class="op">%&gt;%</span> <span class="fu">tail</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 3
  REF   ALT      DP
  &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
1 A     AC        2
2 G     T        10
3 A     G        16
4 A     C        11
5 TGG   T        10
6 A     C         9</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SRR2584863_variants</span> <span class="op">%&gt;%</span> <span class="fu">slice</span><span class="op">(</span><span class="fl">4</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="co"># shows rows 4-10 instead</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7 × 3
  REF                              ALT                                        DP
  &lt;chr&gt;                            &lt;chr&gt;                                   &lt;dbl&gt;
1 CTTTTTTT                         CTTTTTTTT                                  12
2 CCGC                             CCGCGC                                     10
3 C                                T                                          10
4 C                                A                                           8
5 G                                A                                          11
6 ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAG ACAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCCAGCC…     3
7 AT                               ATT                                         7</code></pre>
</div>
<div id="exercise-pipe-and-filter" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise-pipe-and-filter" class="callout-inner">
<h3 class="callout-title">Exercise: Pipe and filter</h3>
<div class="callout-content">
<p>Starting with the <code>variants</code> data frame, use pipes to
subset the data to include only observations from SRR2584863 sample,
where the filtered depth (DP) is at least 10. Showing only 5th through
11th rows of columns <code>REF</code>, <code>ALT</code>, and
<code>POS</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">filter</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"SRR2584863"</span> <span class="op">&amp;</span> <span class="va">DP</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">slice</span><span class="op">(</span><span class="fl">5</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">select</span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">DP</span>, <span class="va">REF</span>, <span class="va">ALT</span>, <span class="va">POS</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7 × 5
  sample_id     DP REF   ALT       POS
  &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 SRR2584863    11 G     A     1733343
2 SRR2584863    20 A     C     2446984
3 SRR2584863    12 G     T     2618472
4 SRR2584863    19 A     T     2665639
5 SRR2584863    15 G     A     2999330
6 SRR2584863    10 A     C     3339313
7 SRR2584863    14 C     A     3401754</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="mutate">Mutate<a class="anchor" aria-label="anchor" href="#mutate"></a></h3>
<p>Frequently you’ll want to create new columns based on the values in
existing columns, for example to do unit conversions or find the ratio
of values in two columns. For this we’ll use the <code>dplyr</code>
function <code>mutate()</code>.</p>
<p>For example, we can convert the polymorphism confidence value QUAL to
a probability value according to the formula:</p>
<p>Probability = 1- 10 ^ -(QUAL/10)</p>
<p>We can use <code>mutate</code> to add a column (<code>POLPROB</code>)
to our <code>variants</code> data frame that shows the probability of a
polymorphism at that site given the data.</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>POLPROB <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="op">(</span><span class="fl">10</span> <span class="op">^</span> <span class="op">-</span><span class="op">(</span><span class="va">QUAL</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 30
   sample_id  CHROM        POS ID    REF   ALT    QUAL FILTER INDEL   IDV    IMF
   &lt;chr&gt;      &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;lgl&gt;  &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt;
 1 SRR2584863 CP000819… 9.97e3 NA    T     G        91 NA     FALSE    NA NA
 2 SRR2584863 CP000819… 2.63e5 NA    G     T        85 NA     FALSE    NA NA
 3 SRR2584863 CP000819… 2.82e5 NA    G     T       217 NA     FALSE    NA NA
 4 SRR2584863 CP000819… 4.33e5 NA    CTTT… CTTT…    64 NA     TRUE     12  1
 5 SRR2584863 CP000819… 4.74e5 NA    CCGC  CCGC…   228 NA     TRUE      9  0.9
 6 SRR2584863 CP000819… 6.49e5 NA    C     T       210 NA     FALSE    NA NA
 7 SRR2584863 CP000819… 1.33e6 NA    C     A       178 NA     FALSE    NA NA
 8 SRR2584863 CP000819… 1.73e6 NA    G     A       225 NA     FALSE    NA NA
 9 SRR2584863 CP000819… 2.10e6 NA    ACAG… ACAG…    56 NA     TRUE      2  0.667
10 SRR2584863 CP000819… 2.33e6 NA    AT    ATT     167 NA     TRUE      7  1
# ℹ 791 more rows
# ℹ 19 more variables: DP &lt;dbl&gt;, VDB &lt;dbl&gt;, RPB &lt;dbl&gt;, MQB &lt;dbl&gt;, BQB &lt;dbl&gt;,
#   MQSB &lt;dbl&gt;, SGB &lt;dbl&gt;, MQ0F &lt;dbl&gt;, ICB &lt;lgl&gt;, HOB &lt;lgl&gt;, AC &lt;dbl&gt;,
#   AN &lt;dbl&gt;, DP4 &lt;chr&gt;, MQ &lt;dbl&gt;, Indiv &lt;chr&gt;, gt_PL &lt;dbl&gt;, gt_GT &lt;dbl&gt;,
#   gt_GT_alleles &lt;chr&gt;, POLPROB &lt;dbl&gt;</code></pre>
</div>
<p>Note, we did not save the new column. We printed the resulting data
frame to the screen. If we want to save the data frame with the new
column, we need to assign it to an object, either overwriting our
<code>variants</code> object (which could be risky of a column of the
same name exists) or creating a new object to store it (which takes up
more space). You will need to think carefully about how to structure
your objects in R.</p>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>There are a lot of columns in our data set, so let’s just look at
the <code>sample_id</code>, <code>POS</code>, <code>QUAL</code>, and
<code>POLPROB</code> columns for now. 1. Add a line to the above code to
only show those columns.</li>
<li>Sometimes we may want to order the result by a column, perhaps to
have the highest values at the top. Look at the help page for
<code>arrange()</code> and see if you can make the highest values at the
top of the resulting data frame. What would you do if you wanted the
lowest values at the top?</li>
</ol></div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" aria-labelledby="headingSolution6" data-bs-parent="#accordionSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">mutate</span><span class="op">(</span>POLPROB <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="fl">10</span> <span class="op">^</span> <span class="op">-</span><span class="op">(</span><span class="va">QUAL</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">select</span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">POS</span>, <span class="va">QUAL</span>, <span class="va">POLPROB</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># part 1</span></span>
<span> <span class="fu">arrange</span><span class="op">(</span><span class="va">POLPROB</span><span class="op">)</span> <span class="co"># part 2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 4
   sample_id      POS  QUAL POLPROB
   &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
 1 SRR2584866 2055833  4.38   0.636
 2 SRR2584866 3538863  4.95   0.680
 3 SRR2584866 3550069  5.05   0.687
 4 SRR2584866 3550071  5.05   0.687
 5 SRR2584866  634603  5.61   0.725
 6 SRR2584866 2055692  5.76   0.734
 7 SRR2584866 1424975  7.45   0.820
 8 SRR2584866  942702  7.91   0.838
 9 SRR2584866 2055662  8.14   0.846
10 SRR2584866 1746738  8.38   0.855
# ℹ 791 more rows</code></pre>
</div>
<p>How can you make the lowest values on the top? Add the
<code>desc()</code> function.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">mutate</span><span class="op">(</span>POLPROB <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="fl">10</span> <span class="op">^</span> <span class="op">-</span><span class="op">(</span><span class="va">QUAL</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">select</span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">POS</span>, <span class="va">QUAL</span>, <span class="va">POLPROB</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span> <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">POLPROB</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 801 × 4
   sample_id      POS  QUAL POLPROB
   &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;
 1 SRR2584863  281923   217       1
 2 SRR2584863  473901   228       1
 3 SRR2584863  648692   210       1
 4 SRR2584863 1331794   178       1
 5 SRR2584863 1733343   225       1
 6 SRR2584863 2333538   167       1
 7 SRR2584863 2446984   225       1
 8 SRR2584863 2665639   225       1
 9 SRR2584863 2999330   225       1
10 SRR2584863 3339313   211       1
# ℹ 791 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="group_by-and-summarize-functions">group_by() and summarize() functions<a class="anchor" aria-label="anchor" href="#group_by-and-summarize-functions"></a></h3>
<p>Many data analysis tasks can be approached using the
“split-apply-combine” paradigm: split the data into groups, apply some
analysis to each group, and then combine the results. <code>dplyr</code>
makes this very easy through the use of the <code>group_by()</code>
function, which splits the data into groups.</p>
<p>We can use <code>group_by()</code> to tally the number of mutations
detected in each sample using the function <code>tally()</code>:</p>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">tally</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 2
  sample_id      n
  &lt;chr&gt;      &lt;int&gt;
1 SRR2584863    25
2 SRR2584866   766
3 SRR2589044    10</code></pre>
</div>
<p>Since counting or tallying values is a common use case for
<code>group_by()</code>, an alternative function was created to bypasses
<code>group_by()</code> using the function <code>count()</code>:</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 2
  sample_id      n
  &lt;chr&gt;      &lt;int&gt;
1 SRR2584863    25
2 SRR2584866   766
3 SRR2589044    10</code></pre>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ul><li>How many mutations are INDELs?</li>
</ul></div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" aria-labelledby="headingSolution7" data-bs-parent="#accordionSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">INDEL</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 2 × 2
  INDEL     n
  &lt;lgl&gt; &lt;int&gt;
1 FALSE   700
2 TRUE    101</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>When the data is grouped, <code>summarize()</code> can be used to
collapse each group into a single-row summary. <code>summarize()</code>
does this by applying an aggregating or summary function to each
group.</p>
<p>It can be a bit tricky at first, but we can imagine physically
splitting the data frame by groups and applying a certain function to
summarize the data.</p>
<center>
<img src="fig/split_apply_combine.png" alt="rstudio default session" style="width: 500px;" class="figure"></center>
<p><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<p>We can also apply many other functions to individual columns to get
other summary statistics. For example,we can use built-in functions like
<code>mean()</code>, <code>median()</code>, <code>min()</code>, and
<code>max()</code>. These are called “built-in functions” because they
come with R and don’t require that you install any additional packages.
By default, all <strong>R functions operating on vectors that contains
missing data will return NA</strong>. It’s a way to make sure that users
know they have missing data, and make a conscious decision on how to
deal with it. When dealing with simple statistics like the mean, the
easiest way to ignore <code>NA</code> (the missing data) is to use
<code>na.rm = TRUE</code> (<code>rm</code> stands for remove).</p>
<p>So to view the mean, median, maximum, and minimum filtered depth
(<code>DP</code>) for each sample:</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span></span>
<span>    mean_DP <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span>,</span>
<span>    median_DP <span class="op">=</span> <span class="fu">median</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span>,</span>
<span>    min_DP <span class="op">=</span> <span class="fu">min</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span>,</span>
<span>    max_DP <span class="op">=</span> <span class="fu">max</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 5
  sample_id  mean_DP median_DP min_DP max_DP
  &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 SRR2584863    10.4      10        2     20
2 SRR2584866    10.6      10        2     79
3 SRR2589044     9.3       9.5      3     16</code></pre>
</div>
<div id="grouped-data-frames-in-tidyverse" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="grouped-data-frames-in-tidyverse" class="callout-inner">
<h3 class="callout-title">Grouped Data Frames in Tidyverse</h3>
<div class="callout-content">
<p>When you group a data frame with <code>group_by()</code>, you get a
grouped data frame. This is a special type of data frame that has all
the properties of a regular data frame but also has an additional
attribute that describes the grouping structure. The primary advantage
of a grouped data frame is that it allows you to work with each group of
observations as if they were a separate data frame.</p>
<p>Operations like <code>summarise()</code> and <code>mutate()</code>
will be applied to each group separately. This is particularly useful
when you want to perform calculations on subsets of your data.</p>
<p>To remove the grouping structure from a grouped data frame, you can
use the <code>ungroup()</code> function. This will return a regular data
frame.</p>
<p>For more details, refer to the <a href="https://dplyr.tidyverse.org/articles/grouping.html" class="external-link">dplyr
documentation on grouping</a>.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="combining-data-frames">Combining data frames<a class="anchor" aria-label="anchor" href="#combining-data-frames"></a></h3>
<p>There are times you may have multiple data tables that have
information that could come together in them. This is often a structure
set up that reflects database management where you you separate out
different information into different related tables. For example, you
may have one table that contains the results from your SNP analysis, as
we have been using in this lesson, and another that has the metadata
about your samples. For some applications you may want to pull
information from that metadata table and be able to use it in your SNP
analysis, maybe particular groupings that are repsented in the metadata
table or alternate names, or other information.</p>
<p>For our example, we want to be able to use the generation number
instead of the <code>sample_id</code> for further analyses and perhaps
plotting later. We have a table of metadata from the <a href="https://datacarpentry.org/organization-genomics/01-tidiness.html" class="external-link">Project
Organization and Management for Genomics</a> lesson that includes
columns that matches up the <code>sample_id</code> column with the
<code>generation</code> information.</p>
<p>First, we will download the metadata table using the
<code>download.file</code> function in R.</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/datacarpentry/wrangling-genomics/gh-pages/files/Ecoli_metadata_composite.tsv"</span>, destfile <span class="op">=</span> <span class="st">"data/Ecoli_metadata_composite.tsv"</span><span class="op">)</span></span></code></pre>
</div>
<p>We will only need to do this once and not every time we run our
script so we can then comment the last line out.</p>
<p>Next we need to load the metadata table into an object in R. This
time we will use the <code>read_delim()</code> function as it has preset
arguments that work well for <em>tab separated value</em> files,
<code>tsv</code>, which is the format this table is</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">read_delim</span><span class="op">(</span><span class="st">"data/Ecoli_metadata_composite.tsv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: One or more parsing issues, call `problems()` on your data frame for details,
e.g.:
  dat &lt;- vroom(...)
  problems(dat)</code></pre>
</div>
<p>This prints a warning message because the last entry is missing some
values but it should still work for our purposes.</p>
<p>Before we match-up our table, we will simplify the metadata table to
only include 3 columns.</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata_sub</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">metadata</span>, <span class="va">strain</span>, <span class="va">generation</span>, <span class="va">run</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we need to consider how we want to merge these two tables
together. There are many ways we might intersect these two tables, for
example: <em>Do we want an intersection that only keeps rows that match
in a certain column from both tables? Do we want to keep only the rows
from one table and the matching information from the other table? Do we
want to keep only the non-matching information from both tables (though
this is less common)?</em></p>
<p>In our case we want to keep all the rows from the
<code>variants</code> table and then pull information that matches from
the <code>metadata_sub</code> table. First, we will explore what happens
when we join these tables in other ways as well as an experiment.</p>
<p>First we will try to keep only the rows, observations, that match in
both tables. This is called an <code>inner_join</code>. For this join we
need to tell it which columns are equivalent in our two data frames. You
will not need to do this if the columns have the same name across data
frames but in our case what is called <code>sample_id</code> in the
<code>variants</code> data frame is called <code>run</code> in the
<code>metadata_sub</code> data frame.</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">inner</span> <span class="op">&lt;-</span> <span class="fu">inner_join</span><span class="op">(</span><span class="va">variants</span>, <span class="va">metadata_sub</span>, by <span class="op">=</span> <span class="fu">join_by</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">run</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>In this case the <code>inner</code> data frame is the same number of
rows (801) as the <code>variants</code> data frame. This is because all
of the <code>sample_id</code>’s that are in the <code>variants</code>
data frame have a corresponding <code>run</code> in
<code>metadata_sub</code>. If one was missing, then all of the data for
that sample would be dropped from the resulting data frame.</p>
<p>We can also see that instead of having 29 columns/variables like
<code>variants</code>, <code>inner</code> has 31 columns/variables. You
may want to run <code>View(inner)</code> and scroll to the far right to
see the new <code>strain</code>, and <code>generation</code> columns
that were added from the <code>metadata_sub</code> table. Note, it did
not add on the <code>run</code> column since that is repeated
information in the <code>sample_id</code> column. Also, we would have
had many more new columns added if we had used the full
<code>metadata</code> data frame instead of the
<code>metadata_sub</code> data frame.</p>
<p>Next we will try an “outer join” which only keeps all observations
which are present in either data frames.</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full</span> <span class="op">&lt;-</span> <span class="fu">full_join</span><span class="op">(</span><span class="va">variants</span>, <span class="va">metadata_sub</span>, by <span class="op">=</span> <span class="fu">join_by</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">run</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>The new <code>full</code> data frame now has more rows than
<code>inner</code> or <code>variants</code>! Though it has the same
number of columns as <code>inner</code> adding on the two additional
<code>strain</code> and <code>generation</code> columns. <em>Why do you
think this has more rows than our original data?</em> Hint: You may want
to <code>View(full)</code> and scroll down to the botton of the data
frame.</p>
<p>Once you look at the bottom of the data frame, you will see a bunch
of <code>sample_id</code>’s that we did not run the SNP analysis on but
were in the <code>metadata_sub</code> data frame. The full join will
keep every unique observation from your <code>by</code> columns even if
they <strong>do not</strong> match up in the other table.</p>
<p>This isn’t quite what we were looking for either.</p>
<p>What we want is to keep all the data in our <code>variants</code>
data frame, even if we do not find a match for it in the
<code>metadata_sub</code> table, so we avoid dropping any data if the
<code>metadata_sub</code> is missing some information. In this case we
will instead use another “outer join” option called “left join”. A left
join keeps all of the data in the table written on the left side of our
function, and will add on columns where the table on the right side
matches via our <code>by</code> statement.</p>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># arguments      "left tbl"  "right tbl"</span></span>
<span><span class="va">left</span> <span class="op">&lt;-</span> <span class="fu">left_join</span><span class="op">(</span><span class="va">variants</span>, <span class="va">metadata_sub</span>, by <span class="op">=</span> <span class="fu">join_by</span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="va">run</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>In this case, the resulting data frame matches our <code>inner</code>
result exactly because there was no missing data in our right table.
Note: A “right” join is the opposite of a “left” so it will keep all the
data in the right most listed table and merge on the info in the left
most listed table</p>
<p>It is important to think carefully about what kind of join you want
and check the resulting data frames to make sure they are what you
expected when you planned your join. In this case, our “inner” and
“left” join are the same but you want to be careful about possibly
dropping or duplicating data depending on how the data frames are
structured.</p>
<p>Finally, we can now do a data manipulation using the
<code>generation</code> column instead of the <code>sample_id</code>
column. We can repeat the calculations for counting SNPs for each
sample</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">left</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">generation</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 2
  generation     n
       &lt;dbl&gt; &lt;int&gt;
1       5000    10
2      15000    25
3      50000   766</code></pre>
</div>
<p>This result makes it easier to see the accumulation of more SNPs at
later generations, without us having to know the sample IDs.</p>
<div id="what-about-right-joins" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="what-about-right-joins" class="callout-inner">
<h3 class="callout-title">What about right joins?</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>How many rows and columns would you expect from the following right
join?</li>
</ol><p><code>right_join(variants, metadata_sub, by = join_by(sample_id == run))</code></p>
<ol start="2" style="list-style-type: decimal"><li>How many rows and columns would you expect from the following right
join?</li>
</ol><p><code>right_join(metadata_sub, variants, by = join_by(run == sample_id))</code></p>
<p>Think carefully about the data in question and which data frame is on
the right and which is on the left.</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8"> Show me the solution </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" aria-labelledby="headingSolution8" data-bs-parent="#accordionSolution8">
<div class="accordion-body">
<p><strong>Part 1</strong> There will be 860 rows and 31 variables, just
like the full join. All of the <code>sample_id</code>’s in the
<code>variants</code> data frame have matches and will be kept and then
it will also add on the <code>run</code> values that do not match but
were represented in the <code>metadata_sub</code> data frame with empty
info in the other columns since there is no matching rows in the
<code>variants</code> data frame.</p>
<p><strong>Part 2</strong> There will be 801 rows and 31 variables, just
like the inner and left joins. This join should always match exactly the
left join as it is the mirrored right join. It will only match the inner
join if all of the samples in the <code>by</code> match-up in the right
data frame are in the left data frame as well, otherwise it will drop
the rows not listed in the left for the inner join.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="reshaping-data-frames---extra">Reshaping data frames - Extra<a class="anchor" aria-label="anchor" href="#reshaping-data-frames---extra"></a></h3>
<p>It can sometimes be useful to transform the “long” tidy format, into
the wide format. This transformation can be done with the
<code>pivot_wider()</code> function provided by the <code>tidyr</code>
package (also part of the <code>tidyverse</code>).</p>
<p><code>pivot_wider()</code> takes a data frame as the first argument,
and two arguments: the column name that will become the columns and the
column name that will become the cells in the wide data.</p>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_wide</span> <span class="op">&lt;-</span> <span class="va">variants</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">CHROM</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarize</span><span class="op">(</span>mean_DP <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">DP</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample_id</span>, values_from <span class="op">=</span> <span class="va">mean_DP</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>`summarise()` has grouped output by 'sample_id'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_wide</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1 × 4
  CHROM      SRR2584863 SRR2584866 SRR2589044
  &lt;chr&gt;           &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 CP000819.1       10.4       10.6        9.3</code></pre>
</div>
<p>The opposite operation of <code>pivot_wider()</code> is taken care by
<code>pivot_longer()</code>. We specify the names of the new columns,
and here add <code>-CHROM</code> as this column shouldn’t be affected by
the reshaping:</p>
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">variants_wide</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span><span class="op">-</span><span class="va">CHROM</span>, names_to <span class="op">=</span> <span class="st">"sample_id"</span>, values_to <span class="op">=</span> <span class="st">"mean_DP"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 3 × 3
  CHROM      sample_id  mean_DP
  &lt;chr&gt;      &lt;chr&gt;        &lt;dbl&gt;
1 CP000819.1 SRR2584863    10.4
2 CP000819.1 SRR2584866    10.6
3 CP000819.1 SRR2589044     9.3</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="resources">Resources<a class="anchor" aria-label="anchor" href="#resources"></a></h3>
<ul><li><p><a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-transformation.pdf" class="external-link">Handy
dplyr cheatsheet</a></p></li>
<li><p><a href="https://github.com/USEPA/workshops/blob/main/r/2015/introR/rmd_posts/2015-01-14-03-Clean.Rmd" class="external-link">Much
of this lesson was copied or adapted from Jeff Hollister’s
materials</a></p></li>
</ul><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>Use the <code>dplyr</code> package to manipulate data frames.</li>
<li>Use <code>glimpse()</code> to quickly look at your data frame.</li>
<li>Use <code>select()</code> to choose variables from a data
frame.</li>
<li>Use <code>filter()</code> to choose data based on values.</li>
<li>Use <code>mutate()</code> to create new variables.</li>
<li>Use <code>group_by()</code> and <code>summarize()</code> to work
with subsets of data.</li>
</ul></div>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr><ol><li id="fn1"><p>The figure was adapted from the Software Carpentry
lesson, <a href="https://swcarpentry.github.io/r-novice-gapminder/13-dplyr/" class="external-link">R for
Reproducible Scientific Analysis</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol></div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="02-data-prelude.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="06-data-visualization.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="02-data-prelude.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to the
        </a>
        <a class="chapter-link float-end" href="06-data-visualization.html" rel="next">
          Next: Data Visualization...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/UW-Madison-DataScience/phmsci756-genomics-r-intro/edit/main/episodes/05-dplyr.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/UW-Madison-DataScience/phmsci756-genomics-r-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/UW-Madison-DataScience/phmsci756-genomics-r-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/UW-Madison-DataScience/phmsci756-genomics-r-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.7" class="external-link">varnish (1.0.7)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://datacarpentry.github.io/genomics-r-intro/05-dplyr.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Data Wrangling and Analyses with Tidyverse",
  "creativeWorkStatus": "active",
  "url": "https://datacarpentry.github.io/genomics-r-intro/05-dplyr.html",
  "identifier": "https://datacarpentry.github.io/genomics-r-intro/05-dplyr.html",
  "dateCreated": "2018-03-12",
  "dateModified": "2025-08-05",
  "datePublished": "2025-08-19"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

